<a href="https://colab.research.google.com/github/thisaraniNJ/Aircraft_Damage_Prediction/blob/main/Aircraft_Damage_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importing Necessary Libraries

In [11]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from warnings import filterwarnings
filterwarnings('ignore')

pd.set_option('display.max_columns', None) # To represent all the columns in the dataset

# Loading Data

In [12]:
Aviation = pd.read_csv('/content/drive/MyDrive/My ML Projects/AviationData.csv', encoding='mac_roman')

In [13]:
Aviation.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude  Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN        NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN        NaN          NaN   
2    Saltville, VA  United States  36.922223 -81.878056          NaN   
3       EUREKA, CA  United States        NaN        NaN          NaN   
4       Canton, OH  United States        NaN        NaN          NaN   

  Airport.Name Injury.Severity Aircraft.damage Aircraft.Category  \
0          NaN        Fatal(2)       Destroyed               NaN   
1          NaN        Fatal(4)       Destroyed               NaN   
2          NaN        Fatal(3)       Destroyed               NaN   
3          NaN        Fatal(2)       Destroyed               NaN   
4          NaN        Fatal(1)       Destroyed               NaN   

  Registration.Number      Make     Model Amateur.Built  Number.of.Engines  \
0              NC6404   Stinson     108-3            No                1.0   
1              N5069P     Piper  PA24-180            No                1.0   
2              N5142R    Cessna      172M            No                1.0   
3              N1168J  Rockwell       112            No                1.0   
4               N15NY    Cessna       501            No                NaN   

     Engine.Type FAR.Description Schedule Purpose.of.flight Air.carrier  \
0  Reciprocating             NaN      NaN          Personal         NaN   
1  Reciprocating             NaN      NaN          Personal         NaN   
2  Reciprocating             NaN      NaN          Personal         NaN   
3  Reciprocating             NaN      NaN          Personal         NaN   
4            NaN             NaN      NaN          Personal         NaN   

   Total.Fatal.Injuries  Total.Serious.Injuries  Total.Minor.Injuries  \
0                   2.0                     0.0                   0.0   
1                   4.0                     0.0                   0.0   
2                   3.0                     NaN                   NaN   
3                   2.0                     0.0                   0.0   
4                   1.0                     2.0                   NaN   

   Total.Uninjured Weather.Condition Broad.phase.of.flight   Report.Status  \
0              0.0               UNK                Cruise  Probable Cause   
1              0.0               UNK               Unknown  Probable Cause   
2              NaN               IMC                Cruise  Probable Cause   
3              0.0               IMC                Cruise  Probable Cause   
4              0.0               VMC              Approach  Probable Cause   

  Publication.Date  
0              NaN  
1       19-09-1996  
2       26-02-2007  
3       12-09-2000  
4       16-04-1980

In [14]:
Aviation.shape

(87282, 31)

There are 87282 flights have been captured to make a prediction

# Data Description

Non-Null Values, Datatypes, % of null values

In [20]:
info_dict = {
    "Column": Aviation.columns.tolist(),
    "Non-Null Count" : Aviation.count(),
    "Dtypes": Aviation.dtypes.tolist(),
    "Null Precentage":(Aviation.shape[0]-Aviation.count())/Aviation.shape[0]*100
}

Description = pd.DataFrame(info_dict).reset_index(drop=True)


In [21]:
Description

Column  Non-Null Count   Dtypes  Null Precentage
0                 Event.Id           87282   object         0.000000
1       Investigation.Type           87282   object         0.000000
2          Accident.Number           87282   object         0.000000
3               Event.Date           87282   object         0.000000
4                 Location           87230   object         0.059577
5                  Country           87056   object         0.258931
6                 Latitude           33073   object        62.107880
7                Longitude           33064   object        62.118192
8             Airport.Code           49217   object        43.611512
9             Airport.Name           51870   object        40.571939
10         Injury.Severity           86399   object         1.011663
11         Aircraft.damage           84272   object         3.448592
12       Aircraft.Category           30783   object        64.731560
13     Registration.Number           85900   object         1.583373
14                    Make           87219   object         0.072180
15                   Model           87190   object         0.105405
16           Amateur.Built           87180   object         0.116863
17       Number.of.Engines           81533  float64         6.586696
18             Engine.Type           81372   object         6.771156
19         FAR.Description           30458   object        65.103916
20                Schedule           12330   object        85.873376
21       Purpose.of.flight           81493   object         6.632524
22             Air.carrier           15971   object        81.701840
23    Total.Fatal.Injuries           75881  float64        13.062258
24  Total.Serious.Injuries           74772  float64        14.332852
25    Total.Minor.Injuries           75349  float64        13.671777
26         Total.Uninjured           81370  float64         6.773447
27       Weather.Condition           83304   object         4.557641
28   Broad.phase.of.flight           61724   object        29.282097
29           Report.Status           82067   object         5.974886
30        Publication.Date           73629   object        15.642400

Statistical Summary

In [22]:
Aviation.describe()

Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       81533.000000          75881.000000            74772.000000   
mean            1.147131              0.652772                0.280921   
std             0.447085              5.521338                1.557476   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            349.000000              161.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          75349.000000     81370.000000  
mean               0.360814         5.311847  
std                2.257721        27.899156  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

In [24]:
Aviation.describe(include='object', exclude='number')

Event.Id Investigation.Type Accident.Number  Event.Date  \
count            87282              87282           87282       87282   
unique           86370                  2           87282       14435   
top     20001212X19172           Accident      SEA87LA080  1984-06-30   
freq                 3              83536               1          25   

             Location        Country Latitude Longitude Airport.Code  \
count           87230          87056    33073     33064        49217   
unique          27171            210    24752     26206        10252   
top     ANCHORAGE, AK  United States  332739N  0112457W         NONE   
freq              434          80998       19        22         1486   

       Airport.Name Injury.Severity Aircraft.damage Aircraft.Category  \
count         51870           86399           84272             30783   
unique        24545             107               4                12   
top         Private       Non-Fatal     Substantial          Airplane   
freq            232           66561           62975             26343   

       Registration.Number    Make  Model Amateur.Built    Engine.Type  \
count                85900   87219  87190         87180          81372   
unique               77696    8045  12185             4             10   
top                   NONE  Cessna    152            No  Reciprocating   
freq                   344   22227   2350         58165          69159   

       FAR.Description Schedule Purpose.of.flight Air.carrier  \
count            30458    12330             81493       15971   
unique              31        3                28       13079   
top                091     NSCH          Personal       Pilot   
freq             17131     4284             48655         246   

       Weather.Condition Broad.phase.of.flight   Report.Status  \
count              83304                 61724           82067   
unique                 4                    12           16666   
top                  VMC               Landing  Probable Cause   
freq               76251                 15428           61754   

       Publication.Date  
count             73629  
unique             2686  
top          25-09-2020  
freq              17019

# Exploratory Data Analysis (EDA)

Data Cleaning

In [26]:
# Check for duplicates
Aviation[Aviation.duplicated()]

Empty DataFrame
Columns: [Event.Id, Investigation.Type, Accident.Number, Event.Date, Location, Country, Latitude, Longitude, Airport.Code, Airport.Name, Injury.Severity, Aircraft.damage, Aircraft.Category, Registration.Number, Make, Model, Amateur.Built, Number.of.Engines, Engine.Type, FAR.Description, Schedule, Purpose.of.flight, Air.carrier, Total.Fatal.Injuries, Total.Serious.Injuries, Total.Minor.Injuries, Total.Uninjured, Weather.Condition, Broad.phase.of.flight, Report.Status, Publication.Date]
Index: []

In [28]:
# Check for missing values
miss_values=Aviation.isnull().sum()
miss_values[miss_values>0]

Location                     52
Country                     226
Latitude                  54209
Longitude                 54218
Airport.Code              38065
Airport.Name              35412
Injury.Severity             883
Aircraft.damage            3010
Aircraft.Category         56499
Registration.Number        1382
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          5749
Engine.Type                5910
FAR.Description           56824
Schedule                  74952
Purpose.of.flight          5789
Air.carrier               71311
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          3978
Broad.phase.of.flight     25558
Report.Status              5215
Publication.Date          13653
dtype: int64

In [29]:
miss_values[miss_values>0].index

Index(['Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [44]:
Null_values = pd.DataFrame() # Create an empty Dataframe to store columns and their null values

Null_values['Columns'] = miss_values[miss_values>0].index.tolist()
Null_values['Null_values']=miss_values[miss_values>0].values.tolist()

# Displaying the percentage of null values for each column and add them to the Null_vlaues dataframe
pct_null_values = list((Aviation.isnull().sum() [Aviation.isnull().sum()>0].values/Aviation.shape[0]*100).round(2))
Null_values['%_null_values']= [str(i)+ '%' for i in pct_null_values]

In [45]:
Null_values

Columns  Null_values %_null_values
0                 Location           52         0.06%
1                  Country          226         0.26%
2                 Latitude        54209        62.11%
3                Longitude        54218        62.12%
4             Airport.Code        38065        43.61%
5             Airport.Name        35412        40.57%
6          Injury.Severity          883         1.01%
7          Aircraft.damage         3010         3.45%
8        Aircraft.Category        56499        64.73%
9      Registration.Number         1382         1.58%
10                    Make           63         0.07%
11                   Model           92         0.11%
12           Amateur.Built          102         0.12%
13       Number.of.Engines         5749         6.59%
14             Engine.Type         5910         6.77%
15         FAR.Description        56824         65.1%
16                Schedule        74952        85.87%
17       Purpose.of.flight         5789         6.63%
18             Air.carrier        71311         81.7%
19    Total.Fatal.Injuries        11401        13.06%
20  Total.Serious.Injuries        12510        14.33%
21    Total.Minor.Injuries        11933        13.67%
22         Total.Uninjured         5912         6.77%
23       Weather.Condition         3978         4.56%
24   Broad.phase.of.flight        25558        29.28%
25           Report.Status         5215         5.97%
26        Publication.Date        13653        15.64%

In [46]:
pct_null_values

[0.06,
 0.26,
 62.11,
 62.12,
 43.61,
 40.57,
 1.01,
 3.45,
 64.73,
 1.58,
 0.07,
 0.11,
 0.12,
 6.59,
 6.77,
 65.1,
 85.87,
 6.63,
 81.7,
 13.06,
 14.33,
 13.67,
 6.77,
 4.56,
 29.28,
 5.97,
 15.64]

Categorical Data

In [47]:
# Using simple Imputer to fill in the null space with the mode values for categorical data
from sklearn.impute import SimpleImputer
Imputer = SimpleImputer(strategy='most_frequent')
Imputer.fit(Aviation.select_dtypes('object'))


SimpleImputer(strategy='most_frequent')

In [51]:
DF = pd.DataFrame(Imputer.transform(Aviation.select_dtypes('object')), columns=Aviation.select_dtypes('object').columns)

In [53]:
DF.isnull().sum()

Event.Id                 0
Investigation.Type       0
Accident.Number          0
Event.Date               0
Location                 0
Country                  0
Latitude                 0
Longitude                0
Airport.Code             0
Airport.Name             0
Injury.Severity          0
Aircraft.damage          0
Aircraft.Category        0
Registration.Number      0
Make                     0
Model                    0
Amateur.Built            0
Engine.Type              0
FAR.Description          0
Schedule                 0
Purpose.of.flight        0
Air.carrier              0
Weather.Condition        0
Broad.phase.of.flight    0
Report.Status            0
Publication.Date         0
dtype: int64

In [54]:
Aviation['Number.of.Engines'].unique()

array([ 1., nan,  2.,  0.,  3.,  4.,  8.])

In [55]:
# reating the number of engines with mode

Aviation['Number.of.Engines'].fillna(Aviation['Number.of.Engines'].mode()[0],inplace=True)

Numerical Data

In [57]:
Aviation.select_dtypes('number').isnull().sum()

Number.of.Engines             0
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
dtype: int64

In [58]:
Imputer_n = SimpleImputer(strategy='mean')
Imputer_n.fit(Aviation.select_dtypes(np.number))
DF_1 = pd.DataFrame(Imputer_n.transform(Aviation.select_dtypes(np.number)), columns=Aviation.select_dtypes(np.number).columns)

In [60]:
DF_1.isnull().sum()

Number.of.Engines         0
Total.Fatal.Injuries      0
Total.Serious.Injuries    0
Total.Minor.Injuries      0
Total.Uninjured           0
dtype: int64

In [61]:
# Concatinating the DF and DF_1

Aviation = pd.concat([DF, DF_1], axis=1)

Aviation

Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
87277  20211223104439           Accident      GAA22WA054  2021-12-21   
87278  20211228104450           Accident      WPR22LA069  2021-12-22   
87279  20211227104443           Accident      CEN22FA082  2021-12-26   
87280  20211230104454           Accident      ERA22LA097  2021-12-29   
87281  20211214104403           Accident      GAA22WA047  2022-12-08   

              Location        Country   Latitude  Longitude Airport.Code  \
0      MOOSE CREEK, ID  United States    332739N   0112457W         NONE   
1       BRIDGEPORT, CA  United States    332739N   0112457W         NONE   
2        Saltville, VA  United States  36.922223 -81.878056         NONE   
3           EUREKA, CA  United States    332739N   0112457W         NONE   
4           Canton, OH  United States    332739N   0112457W         NONE   
...                ...            ...        ...        ...          ...   
87277      Auxerre, OF  United States    475047N   0032948W         NONE   
87278       Auburn, CA  United States    332739N   0112457W         NONE   
87279        Hardy, AR  United States    332739N   0112457W         NONE   
87280     Marathon, FL  United States    332739N   0112457W         NONE   
87281    La Aurora, OF  PACIFIC OCEAN    332739N   0112457W         NONE   

      Airport.Name Injury.Severity Aircraft.damage Aircraft.Category  \
0          Private        Fatal(2)       Destroyed          Airplane   
1          Private        Fatal(4)       Destroyed          Airplane   
2          Private        Fatal(3)       Destroyed          Airplane   
3          Private        Fatal(2)       Destroyed          Airplane   
4          Private        Fatal(1)       Destroyed          Airplane   
...            ...             ...             ...               ...   
87277      Private           Fatal       Destroyed          Airplane   
87278      Private       Non-Fatal     Substantial          Airplane   
87279      Private           Fatal     Substantial          Airplane   
87280      Private       Non-Fatal     Substantial          Airplane   
87281      Private       Non-Fatal     Substantial          Airplane   

      Registration.Number                Make     Model Amateur.Built  \
0                  NC6404             Stinson     108-3            No   
1                  N5069P               Piper  PA24-180            No   
2                  N5142R              Cessna      172M            No   
3                  N1168J            Rockwell       112            No   
4                   N15NY              Cessna       501            No   
...                   ...                 ...       ...           ...   
87277              N565CK  CIRRUS DESIGN CORP      SR22             N   
87278              N6280F              CESSNA      172N             N   
87279              N5798T              CESSNA      172E             N   
87280              N1596U              CESSNA       207             N   
87281              N74521              MOONEY      M20B             N   

         Engine.Type FAR.Description Schedule Purpose.of.flight   Air.carrier  \
0      Reciprocating             091     NSCH          Personal         Pilot   
1      Reciprocating             091     NSCH          Personal         Pilot   
2      Reciprocating             091     NSCH          Personal         Pilot   
3      Reciprocating             091     NSCH          Personal         Pilot   
4      Reciprocating             091     NSCH          Personal         Pilot   
...    

In [62]:
Aviation.isnull().sum()

Event.Id                  0
Investigation.Type        0
Accident.Number           0
Event.Date                0
Location                  0
Country                   0
Latitude                  0
Longitude                 0
Airport.Code              0
Airport.Name              0
Injury.Severity           0
Aircraft.damage           0
Aircraft.Category         0
Registration.Number       0
Make                      0
Model                     0
Amateur.Built             0
Engine.Type               0
FAR.Description           0
Schedule                  0
Purpose.of.flight         0
Air.carrier               0
Weather.Condition         0
Broad.phase.of.flight     0
Report.Status             0
Publication.Date          0
Number.of.Engines         0
Total.Fatal.Injuries      0
Total.Serious.Injuries    0
Total.Minor.Injuries      0
Total.Uninjured           0
dtype: int64